File: JAVASCRIPT/wallet.js
// wallet.js
const { formatDate } = require('./utils');
const Card = require('./card');

class Wallet {
  constructor() {
    this.balance = 0;
    this.actions = [];
    this.idCard = new Card("ID Card", "present", { name: 'John Doe', birthdate: '01/01/1990' });
    this.drivingLicense = new Card("Driving License", "present", { name: 'John Doe', birthdate: '01/01/1990', licence: 'A | B | D' });
  }

  addMoney(amount) {
    this.balance += amount;
    this.actions.push({ type: 'Add Money', amount, balance: this.balance, date: new Date() });
  }

  withdrawMoney(amount) {
    if (amount > this.balance) {
      console.log("Insufficient balance to withdraw this amount.");
      return;
    }

    this.balance -= amount;
    this.actions.push({ type: 'Withdraw Money', amount, balance: this.balance, date: new Date() });
  }

  showBalance() {
    console.log(`Current balance: ${this.balance} MGA`);
  }

  displayHistory() {
    console.log("Action history:");
    this.actions.forEach((action, index) => {
      const formattedDate = formatDate(action.date);
      console.log(`${index + 1}. ${action.type}: ${action.amount} MGA (Balance: ${action.balance} MGA) - Date: ${formattedDate}`);
    });
  }

  displayCardsInfo() {
    this.idCard.displayInfo();
    this.drivingLicense.displayInfo();
  }

  withdrawCard(type) {
    if (type === "ID Card") {
      this.idCard.withdraw();
    } else if (type === "Driving License") {
      this.drivingLicense.withdraw();
    } else {
      console.log("Unrecognized card type.");
    }
  }

  putBackCard(type) {
    if (type === "ID Card") {
      this.idCard.putBack();
    } else if (type === "Driving License") {
      this.drivingLicense.putBack();
    } else {
      console.log("Unrecognized card type.");
    }
  }
}

module.exports = Wallet;

File: PYTHON/wallet.py
from card import Card
from datetime import datetime
from utils import format_date

class Wallet:
    def __init__(self):
        self.balance = 0
        self.actions = []
        self.id_card = Card("ID Card", "present", {"Name": "John Doe", "Birthdate": "01/01/1990"})
        self.driving_license = Card("Driving License", "present", {"Name": "John Doe", "Birthdate": "01/01/1990", "License Number": "ABC123"})


    def add_money(self, amount):
        self.balance += amount
        self.actions.append({'type': 'Add Money', 'amount': amount, 'balance': self.balance, 'date': datetime.now()})

    def withdraw_money(self, amount):
        if amount > self.balance:
            print("Insufficient balance to withdraw this amount.")
            return

        self.balance -= amount
        self.actions.append({'type': 'Withdraw Money', 'amount': amount, 'balance': self.balance, 'date': datetime.now()})

    def show_balance(self):
        print(f"Current balance: {self.balance} MGA")

    def display_history(self):
        print("Action history:")
        for index, action in enumerate(self.actions, start=1):
            formatted_date = format_date(action['date'])
            print(f"{index}. {action['type']}: {action['amount']} MGA (Balance: {action['balance']} MGA) - Date: {formatted_date}")

    def display_cards_info(self):
        self.id_card.display_info()
        self.driving_license.display_info()

    def withdraw_card(self, card_type):
        if card_type == "ID Card":
            self.id_card.withdraw()
        elif card_type == "Driving License":
            self.driving_license.withdraw()
        else:
            print("Unrecognized card type.")

    def put_back_card(self, card_type):
        if card_type == "ID Card":
            self.id_card.put_back()
        elif card_type == "Driving License":
            self.driving_license.put_back()
        else:
            print("Unrecognized card type.")